src/utils.ts

```diff
@@ -138,66 +152,263 @@ export const updateResourceObject = (original: Resource,
+  let newStoreQuery = Object.assign({}, newState[queryId]);
+  newStoreQuery.errors.length = 0;
+  if(document.errors){
+    newStoreQuery.errors.push(...document.errors);
+  }
+  newState[queryId] = newStoreQuery;
+  return newState;
+}
+
+
+export const updateResourceErrors = (state: NgrxJsonApiStoreData, query: ResourceQuery, document : Document): NgrxJsonApiStoreData => {
+  if(!query.type || !query.id || document.data instanceof Array){
+    throw new Error("invalid parameters");
+  }
+  if(!state[query.type] || !state[query.type][query.id]){
+    // resource is not locally stored, no need to update(?)
+    return state;
+  }
+
+  let newState: NgrxJsonApiStoreData = Object.assign({}, state);
+  newState[query.type] = Object.assign({}, newState[query.type]);
+  let storeResource =  Object.assign({}, newState[query.type][query.id]);;
+  storeResource.errors.length = 0;
+  if(document.errors){
+    storeResource.errors.push(...document.errors);
+  }
+  newState[query.type][query.id] = storeResource;
+  return newState;
+}
+
+export const rollbackStoreResources = (state: NgrxJsonApiStoreData): NgrxJsonApiStoreData => {
+  let newState: NgrxJsonApiStoreData = Object.assign({}, state);
+  for(let type in newState){
+    newState[type] = Object.assign({}, newState[type]);
+    for(let id in newState[type]){
+      let storeResource = newState[type][id];
+      if(storeResource.originalResource == null){
+        delete newState[type][id];
+      }else if(storeResource.state != ResourceState.IN_SYNC){
+        newState[type][id] = Object.assign({}, newState[type][id], {
+          state : ResourceState.IN_SYNC,
+          resource :  newState[type][id].originalResource
+        });
+      }
+    }
+  }
+  return newState;
+};
+
 export const updateOrInsertResource = (state: NgrxJsonApiStoreData,
-    resource: Resource): NgrxJsonApiStoreData => {
+    resource: Resource, fromServer : boolean, override : boolean): NgrxJsonApiStoreData => {
 
     let newState: NgrxJsonApiStoreData = Object.assign({}, state);
 
     // handle relationships first.
-    if (resource.hasOwnProperty('relationships')) {
-        Object.keys(resource.relationships)
-            .forEach(relation => {
-                let data = resource.relationships[relation].data;
-                if (_.isPlainObject(data)) {
-                    // hasOne relation
-                    newState = updateOrInsertResource(state, data);
-                } else if (_.isArray(data)) {
-                    // hasMany relation
-                    newState = <NgrxJsonApiStoreData>data.reduce(
-                        (partialState: NgrxJsonApiStoreData, currentResource: Resource): NgrxJsonApiStoreData => {
-                            return updateOrInsertResource(partialState, currentResource);
-                        }, newState);
-                }
-            });
-    }
+    // FIXME this is not working, the data section of a relationship contains only <type, id>, not a complete resource
```

Comment:
What is wrong with this? 

I think what I was trying to do is store the related resource even if it has only a `type` and `id`.